### DISCUSSION TITLE
How to build a statically linked native executable

### DISCUSSION THREAD

**Discussion by rusty-snake** [VirtusLab/scala-cli#3142] [General]
What I'm trying to do:
I want a native app (i.e. `scc --power package [-f] --native [--native-mode=release-fast --native-lto=thin] -o hello hello.sc`) that is statically linked (I only are about Linux, special dependencies on e.g. MacOS are not relevant).

I tried with combinations of `--native-compile="-static"`, `--native-linking="-static"`,`--native-target=static-app`. However they either fail to build or the result is still dynamically linked.

**Is it possible to build a statically linked app?**
If yes, how¹?
If no, is it realistic to get support for it? Should I open a feature request?

---

<details><summary>build error</summary>

```
scala.build.errors.ScalaNativeBuildError: Error compiling with Scala Native
  scala.cli.commands.package0.Package$.buildNative$$anonfun$1(Package.scala:1122)
  scala.build.EitherCps$Helper.apply(EitherCps.scala:19)
  scala.cli.commands.package0.Package$.buildNative(Package.scala:1126)
  scala.cli.commands.package0.Package$.doPackage$$anonfun$1(Package.scala:400)
  scala.build.EitherCps$Helper.apply(EitherCps.scala:19)
  scala.cli.commands.package0.Package$.doPackage(Package.scala:526)
  scala.cli.commands.package0.Package$.runCommand(Package.scala:155)
  scala.cli.commands.package0.Package$.runCommand(Package.scala:71)
  scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:388)
  scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:369)
  caseapp.core.app.CaseApp.main(CaseApp.scala:157)
  scala.cli.commands.ScalaCommand.main(ScalaCommand.scala:354)
  caseapp.core.app.CommandsEntryPoint.main(CommandsEntryPoint.scala:169)
  scala.cli.ScalaCliCommands.main(ScalaCliCommands.scala:125)
  scala.cli.ScalaCli$.main0(ScalaCli.scala:296)
  scala.cli.ScalaCli$.main(ScalaCli.scala:119)
  scala.cli.ScalaCli.main(ScalaCli.scala)
```

</details>

¹ W/o creating a dedicated musl/alpine build environment.

**Comment by WojciechMazur** [comment]
I'm not sure why it might not be working for you. In general it should work, but would be depending on the used libraries and environment on which we're building. Here's one example from gitpod (linux) env: 

```bash
 scala --power package -f --native -o hello sandbox/src/main/scala/Test.scala --native-linking=-static --native-compile=-static
Compiling project (Scala 3.5.0, Scala Native 0.5.4)
Compiled project (Scala 3.5.0, Scala Native 0.5.4)
...
[info] Linking with [pthread, dl]
[error] clang++: warning: argument unused during compilation: '-rdynamic' [-Wunused-command-line-argument]
[info] Linking native code (immix gc, none lto) (212 ms)
[info] Postprocessing (0 ms)
[info] Total (12614 ms)
Wrote /workspace/scala-native/hello, run it with
  ./hello
gitpod /workspace/scala-native (main) $ ./hello 
Hello, World!
gitpod /workspace/scala-native (main) $ ldd ./hello 
        not a dynamic executable
gitpod /workspace/scala-native (main) $ file ./hello 
./hello: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=23276673357367558b823759e485502ed9fbf239, for GNU/Linux 3.2.0, not stripped
```

I'm aware of [issues on MacOS](https://stackoverflow.com/questions/3801011/ld-library-not-found-for-lcrt0-o-on-osx-10-6-with-gcc-clang-static-flag) when compiling with `-static` but it is not Scala Native specific. 

I believe their logs are missing something that would hint why the build failed. For example, here's what I get when building on MacOS: 
```
[error] ld: library not found for -lcrt0.o
[error] clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Reply by rusty-snake** [reply]
Ok It had problems finding all required `lib*.a`. Together with rule 1. "Linking errors are the kind of errors you don't want to happen if you're the guy who ends up debugging them" and rule 2. "Scala is the programming language with the most incomplete documentation [in my toolbox]" this makes you stop early and ask if it is even possible.

However there is one thing I do not understand yet, why does `ldd` say `not a dynamic executable`? For my Rust musl binaries it tells me `statically linked`. Aren't they the same? Have to do some research.

**Reply by rusty-snake** [reply]
Ok, full working command:

```bash
sudo dnf install glibc-static libstdc++-static
scc --power package --native --native-mode=release-fast \
    --native-lto=thin --native-compile="-static-pie" --native-linking="-static-pie" -o hello hello.sc
strip hello
```
